docs/providers/page-api-version.tsx

```diff
@@ -0,0 +1,64 @@
+import { NextRouter } from 'next/router';
+import React, { createContext, PropsWithChildren, useCallback, useContext } from 'react';
+
+export const PageApiVersionContext = createContext({
+  /** The version selected in the URL, or the default version */
+  version: 'latest',
+  /** If the current URL has a version defined */
+  hasVersion: false,
+  /** Change the URL to the selecte version  */
+  setVersion: (newVersion: string): void => {
+    throw new Error('PageApiVersionContext not found');
+  },
+});
+
+type Props = PropsWithChildren<{
+  /** The router containing the current URL info of the page, possibly containing the API version */
+  router: NextRouter;
+}>;
+
+export function PageApiVersionProvider(props: Props) {
+  const version = getVersionFromPath(props.router.pathname);
+  const hasVersion = version !== null;
+
+  // note: if the page doesn't exists, the error page will handle it
+  const setVersion = useCallback((newVersion: string) => {
+    props.router.push(replaceVersionInPath(props.router.pathname, newVersion));
+  }, []);
+
+  return (
+    <PageApiVersionContext.Provider
+      value={{ setVersion, hasVersion, version: version || 'latest' }}>
+      {props.children}
+    </PageApiVersionContext.Provider>
+  );
+}
+
+export function usePageApiVersion() {
+  return useContext(PageApiVersionContext);
+}
+
+/**
+ * Determine if there is a version within the pathname of the URL.
+ * Versioned pages always starts with /versions/<version>.
+ */
+export function isVersionedPath(path: string) {
+  return path.startsWith('/versions/');
+}
+
+/**
+ * Find the version within the pathname of the URL.
+ * This only accepts pathnames, without hashes or query strings.
+ */
+export function getVersionFromPath(path: string) {
+  return !isVersionedPath(path) ? null : path.split('/', 3).pop()!;
```

Comment:
```suggestion
  const versionCaptureRegex = /\/versions\/v((\d+|\.)*)?/
  return !isVersionedPath(path) ? null : path.match(versionCaptureRegex)[1];
```

Just another way to write this that may be a little clearer. I'm hoping we could add a variable to make this more readable.

Reply:
I see what you mean, but the unfortunate thing here is that these values are all valid versions:
- `v44.0.0` -> captured well by this regex
- `v41.0.0` -> same, well captured
- `latest` -> not captured
- `unversioned` -> not captured

To create a regex out of that, I think this might be a bit easier 🙈  Definitely ok with swapping this for regexes, maybe a good follow-up PR?